File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ReportingHelper.java
package org.jellyfin.androidtv.util.apiclient;

import androidx.annotation.Nullable;

import org.jellyfin.androidtv.data.compat.StreamInfo;
import org.jellyfin.androidtv.data.model.DataRefreshService;
import org.jellyfin.androidtv.ui.playback.PlaybackController;
import org.jellyfin.androidtv.ui.playback.PlaybackManager;
import org.jellyfin.apiclient.interaction.ApiClient;
import org.jellyfin.apiclient.interaction.EmptyResponse;
import org.jellyfin.apiclient.model.session.PlaybackProgressInfo;
import org.jellyfin.apiclient.model.session.PlaybackStartInfo;
import org.jellyfin.apiclient.model.session.PlaybackStopInfo;
import org.jellyfin.sdk.model.api.BaseItemKind;
import org.koin.java.KoinJavaComponent;

import java.time.Instant;

import kotlin.Lazy;
import timber.log.Timber;

public class ReportingHelper {
    private final Lazy<PlaybackManager> playbackManager = KoinJavaComponent.<PlaybackManager>inject(PlaybackManager.class);
    private final Lazy<DataRefreshService> dataRefreshService = KoinJavaComponent.<DataRefreshService>inject(DataRefreshService.class);
    private final Lazy<ApiClient> apiClient = KoinJavaComponent.<ApiClient>inject(ApiClient.class);

    public void reportStopped(org.jellyfin.sdk.model.api.BaseItemDto item, StreamInfo streamInfo, long pos) {
        if (item != null && streamInfo != null) {
            Timber.i("ReportingHelper.reportStopped called for " + item.getId() + " at position " + pos);
            PlaybackStopInfo info = new PlaybackStopInfo();
            info.setItemId(item.getId().toString());
            info.setPositionTicks(pos);
            playbackManager.getValue().reportPlaybackStopped(info, streamInfo, apiClient.getValue(), new EmptyResponse());

            dataRefreshService.getValue().setLastPlayback(Instant.now());
            switch (item.getType()) {
                case MOVIE:
                    dataRefreshService.getValue().setLastMoviePlayback(Instant.now());
                    break;
                case EPISODE:
                    dataRefreshService.getValue().setLastTvPlayback(Instant.now());
                    break;
            }
        }
    }

    public void reportStart(org.jellyfin.sdk.model.api.BaseItemDto item, long pos) {
        PlaybackStartInfo startInfo = new PlaybackStartInfo();
        startInfo.setItemId(item.getId().toString());
        if (item.getType() != BaseItemKind.TV_CHANNEL) startInfo.setPositionTicks(pos);
        playbackManager.getValue().reportPlaybackStart(startInfo, apiClient.getValue(), new EmptyResponse());
        Timber.i("Playback of %s started.", item.getName());
    }

    public void reportProgress(@Nullable PlaybackController playbackController, org.jellyfin.sdk.model.api.BaseItemDto item, StreamInfo currentStreamInfo, Long position, boolean isPaused) {
        if (item != null && currentStreamInfo != null) {
            PlaybackProgressInfo info = new PlaybackProgressInfo();
            info.setItemId(item.getId().toString());
            if (item.getType() != BaseItemKind.TV_CHANNEL) {
                info.setPositionTicks(position);
                info.setCanSeek(currentStreamInfo.getRunTimeTicks() != null && currentStreamInfo.getRunTimeTicks() > 0);
            }
            info.setIsPaused(isPaused);
            info.setPlayMethod(currentStreamInfo.getPlayMethod());
            if (playbackController != null && playbackController.isPlaying()) {
                info.setAudioStreamIndex(playbackController.getAudioStreamIndex());
                info.setSubtitleStreamIndex(playbackController.getSubtitleStreamIndex());
            }
            playbackManager.getValue().reportPlaybackProgress(info, currentStreamInfo, apiClient.getValue(), new EmptyResponse());
        }
    }
}

File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ReportingHelper.kt
package org.jellyfin.androidtv.util.apiclient

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.launch
import org.jellyfin.androidtv.data.compat.StreamInfo
import org.jellyfin.androidtv.data.model.DataRefreshService
import org.jellyfin.androidtv.ui.playback.PlaybackController
import org.jellyfin.sdk.api.client.ApiClient
import org.jellyfin.sdk.api.client.extensions.playStateApi
import org.jellyfin.sdk.model.api.BaseItemDto
import org.jellyfin.sdk.model.api.BaseItemKind
import org.jellyfin.sdk.model.api.PlaybackOrder
import org.jellyfin.sdk.model.api.PlaybackProgressInfo
import org.jellyfin.sdk.model.api.PlaybackStartInfo
import org.jellyfin.sdk.model.api.PlaybackStopInfo
import org.jellyfin.sdk.model.api.RepeatMode
import timber.log.Timber
import java.time.Instant

class ReportingHelper(
	private val dataRefreshService: DataRefreshService,
	private val api: ApiClient,
) {
	fun reportStart(
		lifecycleOwner: LifecycleOwner,
		playbackController: PlaybackController?,
		item: BaseItemDto,
		streamInfo: StreamInfo,
		position: Long,
		paused: Boolean
	) {
		val info = PlaybackStartInfo(
			itemId = item.id,
			positionTicks = position,
			canSeek = (streamInfo.runTimeTicks ?: 0) > 0,
			isPaused = paused,
			liveStreamId = streamInfo.mediaSource?.liveStreamId,
			playSessionId = streamInfo.playSessionId,
			playMethod = streamInfo.playMethod,
			audioStreamIndex = playbackController?.audioStreamIndex,
			subtitleStreamIndex = playbackController?.subtitleStreamIndex,
			isMuted = false,
			repeatMode = RepeatMode.REPEAT_NONE,
			playbackOrder = PlaybackOrder.DEFAULT,
			mediaSourceId = streamInfo.mediaSourceId,
		)

		lifecycleOwner.lifecycleScope.launch(Dispatchers.IO + NonCancellable) {
			Timber.i("Reporting ${item.name} playback started at $position")
			runCatching {
				api.playStateApi.reportPlaybackStart(info)
			}.onFailure { error -> Timber.e(error, "Failed to report started playback!") }
		}
	}

	fun reportProgress(
		lifecycleOwner: LifecycleOwner,
		playbackController: PlaybackController?,
		item: BaseItemDto,
		streamInfo: StreamInfo,
		position: Long,
		paused: Boolean
	) {
		val info = PlaybackProgressInfo(
			itemId = item.id,
			positionTicks = position,
			canSeek = (streamInfo.runTimeTicks ?: 0) > 0,
			isPaused = paused,
			liveStreamId = streamInfo.mediaSource?.liveStreamId,
			playSessionId = streamInfo.playSessionId,
			playMethod = streamInfo.playMethod,
			audioStreamIndex = playbackController?.audioStreamIndex,
			subtitleStreamIndex = playbackController?.subtitleStreamIndex,
			isMuted = false,
			repeatMode = RepeatMode.REPEAT_NONE,
			playbackOrder = PlaybackOrder.DEFAULT,
			mediaSourceId = streamInfo.mediaSourceId,
		)

		lifecycleOwner.lifecycleScope.launch(Dispatchers.IO + NonCancellable) {
			Timber.d("Reporting ${item.name} playback progress at $position")
			runCatching {
				api.playStateApi.reportPlaybackProgress(info)
			}.onFailure { error -> Timber.w(error, "Failed to report playback progress") }
		}
	}

	fun reportStopped(lifecycleOwner: LifecycleOwner, item: BaseItemDto, streamInfo: StreamInfo, position: Long?) {
		val info = PlaybackStopInfo(
			itemId = item.id,
			positionTicks = position,
			mediaSourceId = streamInfo.mediaSourceId,
			liveStreamId = streamInfo.mediaSource?.liveStreamId,
			playSessionId = streamInfo.playSessionId,
			failed = false,
		)

		lifecycleOwner.lifecycleScope.launch(Dispatchers.IO + NonCancellable) {
			Timber.i("Reporting ${item.name} playback stopped at $position")
			runCatching {
				api.playStateApi.reportPlaybackStopped(info)
			}.onFailure { error -> Timber.e(error, "Failed to report stopped playback!") }
		}

		// Update dataRefreshService
		dataRefreshService.lastPlayback = Instant.now()
		when (item.type) {
			BaseItemKind.MOVIE -> dataRefreshService.lastMoviePlayback = Instant.now()
			BaseItemKind.EPISODE -> dataRefreshService.lastTvPlayback = Instant.now()
			else -> Unit
		}
	}
}
